models/User.go

```diff
@@ -949,6 +969,137 @@ func CancelOrder(userId uint32, orderId uint32, isAsk bool) (*Ask, *Bid, error)
 	}
 }
 
+// Helper function to determine what percentage the user should be taxed
+func getTaxPercent(netCash int64) uint64 {
+
+	keys := make([]int64, 0)
+	for taxBracket := range TaxBrackets {
+		keys = append(keys, taxBracket)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
+
+	for _, taxBracket := range keys {
+		if netCash <= taxBracket {
+			return TaxBrackets[taxBracket]
+		}
+	}
+	return MaxTaxPercent
+}
+
+// Helper to calculate tax for bidding user
+func getTaxForBiddingUser(tx *gorm.DB, userId uint32, stockId uint32, l *logrus.Entry, stockQuantity uint64, stockPrice uint64, biddingUser *User) *TransactionSummary {
```

Comment:
do you need to pass userId and user separately?

Reply:
Don't pass logrus.Entry. Let this function use its own.

Also add a comment that the User's lock is assumed to have been already acquired, since you're modifying user's cash. Secondly, if this/commit-of-txSummary fails, are you setting user's cash back to original value?

Reply:
Avoid having so many type conversions throughout the code. Either create variables of correct type (specifically talking about the float64 conversions throughout the code) at the top of the function and use them, or use integers everywhere.

Reply:
Can't this and getTaxForAskingUser be combined into same? Haven't looked at the code in detail, so if not possible, ignore this comment.

Reply:
Won't it be simpler to update the db in this function itself (and rename the function appropriately?). If there's error while saving, revert change to user's cash, and return. Let caller rollback the tx.

Reply:
1. userID no longer passed.
2. logrus no longer passed. Comment added. The function itself can't fail because it's only performing arithmetic operations. When the commit actually happens though, there is a rollback in place if it fails.
3. I've reduced the number of type conversions as you suggested.
4. No they shouldn't be combined. It really complicates things more than it would simplify.
5. I'm keeping the db updation in the main function because that's where all other db updation steps are happening. Seems weird to keep the db updation for tax calculation alone in a separate function.